### DISCUSSION TITLE
Error occurred prerendering page "/404"

### DISCUSSION THREAD

**Discussion by eiskalteschatten** [vercel/next.js#12246] [Help] [answered]
I keep getting the following error when running `next build` on my project:

```
Automatically optimizing pages .
Error occurred prerendering page "/404". Read more: https://err.sh/next.js/prerender-error:
Error: Error for page /_error: pages with `getServerSideProps` can not be exported. See more info here: https://err.sh/next.js/gssp-export
```

I don't have a custom 404 page because I want it to be handled by the `_error.tsx` page since I have server-side redirecting for trailing slashes. This works when running in a development environment, but dies when I try to build it.

Obviously `_error.tsx` has `getServerSideProps` and should NOT be a static page, so why is it trying to make it one? It is obviously not meant to be one as according to the Next.js documentation, any page that exports `getServerSideProps` will not be one. So why is it throwing an error?!?!?!

If it helps, here's the code for my `_error.tsx` file:

```
import React, { useEffect } from 'react';
import { GetServerSideProps } from 'next';
import Head from 'next/head';
import Router from 'next/router';

import { makeStyles, createStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';

import useStandardHeaderTags from '../lib/useStandardHeaderTags';
import TitleElement from '../components/TitleElement';

const useStyles = makeStyles(() =>
  createStyles({
    root: {
      textAlign: 'center'
    }
  })
);

interface Props {
  statusCode: number;
}

const Error: React.FC<Props> = ({ statusCode }) => {
  const classes = useStyles();
  const title = statusCode === 404 ? '404' : 'Error';

  return (
    <>
      <Head>
        {useStandardHeaderTags(title)}
      </Head>
      <Container className={classes.root}>
        <TitleElement text={title} />

        {statusCode === 404
          ? 'The page you are looking for could not be found.'
          : 'An error occurred.'}
      </Container>
    </>
  );
};

export const getServerSideProps: GetServerSideProps = async ({ res, req }) => {
  const statusCode = res ? res.statusCode : 404;

  if (statusCode === 404) {
    if (req.url.match(/\/$/)) {
      const withoutTrailingSlash = req.url.substr(0, req.url.length - 1);
      if (res) {
        res.writeHead(303, {
          Location: withoutTrailingSlash
        });
        res.end();
      }
      else {
        Router.push(withoutTrailingSlash);
      }
    }
  }

  return {
    props: {
      statusCode
    }
  };
};

export default Error;
```

**Comment by eiskalteschatten** [comment]
As a follow-up: I also saw this discussion https://github.com/zeit/next.js/discussions/11945, but the answer doesn't help me in anyway because I do not want the error page to be static.

**Comment by eiskalteschatten** [comment]
This can't be proper behavior. I created an issue for it: https://github.com/zeit/next.js/issues/12256

**Comment by eiskalteschatten** [comment, marked as answer]
I was able to solve it by changing the `getServierSideProps` function to

```
Error.getInitialProps = ({ res, req, err }): Props => {
  const statusCode = res ? res.statusCode : err ? err.statusCode : 404;

  if (statusCode === 404) {
    if (req.url.match(/\/$/)) {
      const withoutTrailingSlash = req.url.substr(0, req.url.length - 1);
      if (res) {
        res.writeHead(303, {
          Location: withoutTrailingSlash
        });
        res.end();
      }
      else {
        Router.push(withoutTrailingSlash);
      }
    }
  }

  return { statusCode };
};
```

**Comment by svassaux** [comment]
same issue here, can't solve it easily with initial props

I have the same need because I am localizing my 404 page

works fine otherwise, a 404 or a 500 is handled by the error.tsx, I don't see the point of trying to generate a 404 static page

**Reply by nfantone** [reply]
In recent updates, this also happens with `/500`.

**Reply by chiranjivDev** [reply]
any updates on this?

**Comment by quant-daddy** [comment]
Any updates on this? I'm getting the same error for /500 page

**Reply by quant-daddy** [reply]
Okay so this solves it: I added the `_error.tsx` page in the root pages directory: 
```typescript
import Error from 'next/error';


function Page({ statusCode }) {
  return <Error statusCode={statusCode}></Error>;
}

Page.getInitialProps = ({ res, err }) => {
  const statusCode = res ? res.statusCode : err ? err.statusCode : 404;
  return { statusCode };
};

export default Page;
```

**Reply by cgstaber** [reply]
I had this problem as well and adding this page fixed it! 
Thanks!

**Reply by Laurelin** [reply]
I also had a prerender error on /404 and /500.

This fixed it.

**Reply by Nuruzzamancse** [reply]
Thanks it's solved the problem.

**Reply by 3pns** [reply]
thank you for the fix. I had the same error with the default error pages on next 11.1.0

**Reply by hamen** [reply]
It solved it for 400/500, but now I have it on `/` 🤣

```
Error occurred prerendering page "/". Read more: https://nextjs.org/docs/messages/prerender-error
```
🤦

**Reply by wsasouza** [reply]
I managed to solve it with this tip! Thank you very much!

**Reply by LuisValgoi** [reply]
@wsasouza were you using app router?

**Reply by ohabash** [reply]
![Screenshot 2024-12-30 at 3 13 06 PM](https://github.com/user-attachments/assets/3d2837af-7bc8-499a-97fa-91cf402ac241)
it didnt work when i made `(pages)/error.tsx` nor `(pages)/_error/404/page.tsx`

**Comment by nemanjam** [comment]
I have similar error on prod build. I don't like the fact that Next.js doesn't report any errors during dev and then when you try to compile for prod a lot of stuff breaks. If prod doesn't compile dev shouldn't compile too.

**Comment by LuisValgoi** [comment]
I am facing the same issue when in app router. I have a `src/app/dashboard/page.tsx` that when I try to `throw new Error()`, it gives `Error occurred prerendering page`.

**Reply by mentamarindos** [reply]
same issue, i also have a `dashboard/page.tsx` and a `middleware.ts` to guard that page with auth

**Comment by YChebotaev** [comment]
I have similar error, but I don't even have error.tsx or any pages except for layout.tsx. App router, next 15.1.7

**Comment by glneto** [comment]
The fix in my case was downgrading React to version 18 (was using 19)

**Comment by danidb** [comment]
I had this issue in a monorepo, with npm, nextjs 15.1.7 and 15.2 and react/react-dom ^19; dev worked but build failed. 

I was able to resolve this using `npm dedup` and updating packages that wanted an older versions of react. 

`npm dedup`, caught that I had two "unresolvable" versions of react/react-dom (19.0.0 and 18.something). In my case, the packages that wanted the older version of react were "vaul" and "next-themes,"; they both support react 19 now. If you're using shadcn/ui you probably also depend on these packages.

**Reply by VulcanTech-io** [reply]
i had this same problem, thanks for your help

**Reply by jacob-wt** [reply]
holy sh*t `npm dedupe` fixed it 😭 🎉 thank you!

**Reply by VulcanTech-io** [reply]
I ran into this problem again upgrading to 15.3.1 downgrading to 15.2.4 and then upgrading again fixed it which is just some black magic f... But in case this helps anyone else thought id post about it.

**Reply by quangthien27** [reply]
`npm dedupe` fixed it for me too! Make sure react and react-dom have exact same version too

**Comment by ranaivomahaleo** [comment]
In my case (`next 15.2.4, react 18.3, react-dom 18.3`), this error was related to the environment. When using `next build` be sure that the value of `NODE_ENV` is set to `production` or an empty string, otherwise, this prerendering error is appearing.

The simplest way to verify the value of `NODE_ENV` is using the following command line:

`echo $NODE_ENV`

If the above command returns an empty string or a string equals to `production`, the `next build` can produce the normal behaviour building the next code.
If the above command returns a string which is not equal to `production`, the prerendering error occured.

**Reply by theboringhumane** [reply]
This works.

**Comment by Johanfeliz08** [comment]
In my case i had the optimizedCss option active on my next.config.ts, i fixed just by removing it.

On next.config.ts, remove: 

  experimental: {
    optimizeCss: true,
  }
